File: 0290-word-pattern/0290-word-pattern.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        words=s.split()
        if len(words)!=len(pattern):
            return False
        
        patttoword={}
        wordtopatt={}
        for char,wd in zip(pattern,words):
            if char in patttoword and patttoword[char]!=wd:
                return False
            if wd in wordtopatt and wordtopatt[wd]!=char:
                return False
            patttoword[char]=wd
            wordtopatt[wd]=char
        return True         

File: 0290-word-pattern/0290-word-pattern.java
class Solution {
    public boolean wordPattern(String pattern, String s) {
        String[] words = s.split(" ");
        if (words.length != pattern.length()) {
            return false;
        }
        HashMap<Character, String> pattToWord = new HashMap<>();
        HashMap<String, Character> wordToPatt = new HashMap<>();
        for (int i = 0; i < pattern.length(); i++) {
            char ch = pattern.charAt(i);
            String wd = words[i];
            if (pattToWord.containsKey(ch) && !pattToWord.get(ch).equals(wd)) {
                return false;
            }
            if (wordToPatt.containsKey(wd) && !wordToPatt.get(wd).equals(ch)) {
                return false;
            }
            pattToWord.put(ch, wd);
            wordToPatt.put(wd, ch);
        }
        return true;
    }
}